### ISSUE TITLE
Reduce compile time by removing unnecessary headers

### ISSUE THREAD

**Issue by rcurtin** [mlpack/mlpack#722]
The compile time for mlpack is very long compared to other libraries of the same size.  There are many reasons for this; Armadillo (which is basically header-only) is one; boost is another; the template metaprogramming inside of mlpack is yet another.  But there are some ways that we can reduce compile time.

 * Some of the headers in `src/mlpack/core/prereqs.hpp` are not used in all parts of the project and can be removed (or in some cases, included only where they are necessary).
 * Where possible, we can move header files into .cpp files (so they are not included by everything).  Some examples here might be the boost math headers.
 * We may be able to use `extern template` definitions for some of the various parts mlpack provides, specifying to common types, e.g., `extern template AllkNN<arma::mat, EuclideanDistance, KDTree>` and then having that specialization declared in some .cpp file... or something like this.

It'll be easy to check if you've reduced compile times when Travis builds a PR; right now Travis takes about 31 minutes for a complete build, and AppVeyor takes about 45 minutes.  I doubt we can get order-of-magnitude compile speedups or anything, but we can probably help out incrementally.

**Comment by anshuman23** [comment]
Hi Ryan, I'd like to try my hand at this. I just have a trivial question, and I hope it isn't too kiddish:
- For the last point, do you mean that I should look for similar/repetitive definitions in mlpack and hence implement ```extern template``` definitions for them where I find them repeating? 

As for the first 2 points, I've been looking at the source code and I'll try my best to redistribute the headers so as to minimize compile time.

**Comment by rcurtin** [comment]
Great, I look forward to incorporating some changes to speed things up. :)

You're right about the last point, that is indeed what I meant.  Another nice thing is that if we are providing instantiated templates for common mlpack objects, then users who compile things against mlpack downstream should have faster compile times.

**Comment by anshuman23** [comment]
Alright, thank you! :D
Also, if I have to check for a reduction in compile time, would I have to compile from source every time to check?

**Comment by rcurtin** [comment]
Yeah, if you wanted to check, you would need to `make clean` then `make` (or `make -jX` if you want to build with X cores).  Alternately, you could open a PR and push changes to it, and Travis will build it, but Travis will take about 30-45 minutes per build and there will be a delay before the build starts, so that may not be the most efficient way to do it.

**Comment by sbrodehl** [comment]
@rcurtin You mentioned ```src/mlpack/prereqs.hpp``` is not needed in all parts and can be removed if this is the case, but how about ``` src/mlpack/core.hpp```? I think not all headers there are used in every part of the code, too, right?
Can one split up the headers there? Or use precompiled headers?
These kind of things should speed up the build process.

Or is there a special reason why those headers combine almost all other headers?

**Comment by rcurtin** [comment]
I would not be opposed to pruning `core.hpp`. It was originally made for convenience, so that you could include basically all of the mlpack core, but if we are trying to accelerate compile time we should probably change the rest of the codebase to only include what is needed. I think we should still leave core.hpp there for users, though.

**Comment by Erick2206** [comment]
@rcurtin anyone still working on this issue? Looking forward to start contributing with this issue.

**Comment by rcurtin** [comment]
@Erick2206: you are more than welcome to contribute.  This is a multifaceted issue so any improvement from whomever is welcome.

**Comment by hitesh1157** [comment]
I want to work on this issue. I m facing a problem though. $ cmake shows error !?

**Comment by jayantsaraf** [comment]
@rcurtin can you please give some examples of the header that i need to remove?

**Comment by rcurtin** [comment]
@jayantsaraf: you will have to look through the code to see which headers are needed and which are not.  You can also add `-H` to your `CMAKE_CXX_FLAGS` and this will print a (very long) list of all files that are being included for each source file.

**Comment by omkar13** [comment]
Is this issue still open? I would like to start contributing.

**Comment by zoq** [comment]
@omkar13  Yes, this issue is pretty much open ended. Recently I read an interesting blog post about the ``d2cgsummary`` (http://aras-p.info/blog/2017/10/23/Best-unknown-MSVC-flag-d2cgsummary/) flag which might be helpful or the "Templight" profiling tool (https://github.com/mikael-s-persson/templight) might also come in handy.

**Comment by kaushal07wick** [comment]
hello, I would like to help fixing these problem  Is this still open. thanks
